src/core/search/base.h

```diff
@@ -79,16 +79,16 @@ struct DocumentAccessor {
 //
 // Queries should be done directly on subclasses with their distinc
 // query functions. All results for all index types should be sorted.
-struct BaseIndex {
+template <typename T> struct BaseIndex {
```

Comment:
If you do all of this because you want to use the hnsw index with different types, it seems like it would be easier to branch off somewhere there. Maybe create a templated "base hnsw index" type and a subtype that implements BaseIndex (if we need it at all). Either way it looks simpler than changing everythign up the chain